perses/tests/test_topology_proposal.py

```diff
@@ -0,0 +1,440 @@
+        return 'pdbx'
+    if '.pdb' in filename:
+        return 'pdb'
+    for line in file:
+        if line.startswith('data_') or line.startswith('loop_'):
+            file.seek(0)
+            return 'pdbx'
+        if line.startswith('HEADER') or line.startswith('REMARK') or line.startswith('TITLE '):
+            file.seek(0)
+            return 'pdb'
+    # It's certainly not a valid PDBx/mmCIF.  Guess that it's a PDB.
+    file.seek(0)
+    return 'pdb'
+
+def _matchResidue(res, template, bondedToAtom):
+    atoms = list(res.atoms())
+    if len(atoms) != len(template.atoms):
+        return None
+    matches = len(atoms)*[0]
+    hasMatch = len(atoms)*[False]
+
+    renumberAtoms = {}
+    for i in range(len(atoms)):
+        renumberAtoms[atoms[i].index] = i
+    bondedTo = []
+    externalBonds = []
+    for atom in atoms:
+        bonds = [renumberAtoms[x] for x in bondedToAtom[atom.index] if x in renumberAtoms]
+        bondedTo.append(bonds)
+        externalBonds.append(len([x for x in bondedToAtom[atom.index] if x not in renumberAtoms]))
+
+    residueTypeCount = {}
+    for i, atom in enumerate(atoms):
+        key = (atom.element, len(bondedTo[i]), externalBonds[i])
+        if key not in residueTypeCount:
+            residueTypeCount[key] = 1
+        residueTypeCount[key] += 1
+    templateTypeCount = {}
+    for i, atom in enumerate(template.atoms):
+        key = (atom.element, len(atom.bondedTo), atom.externalBonds)
+        if key not in templateTypeCount:
+            templateTypeCount[key] = 1
+        templateTypeCount[key] += 1
+    if residueTypeCount != templateTypeCount:
+        return None
+
+    if _findAtomMatches(atoms, template, bondedTo, externalBonds, matches, hasMatch, 0):
+        return matches
+    return None
+
+
+def _findAtomMatches(atoms, template, bondedTo, externalBonds, matches, hasMatch, position):
+    if position == len(atoms):
+        return True
+    elem = atoms[position].element
+    name = atoms[position].name
+    for i in range(len(atoms)):
+        atom = template.atoms[i]
+        if ((atom.element is not None and atom.element == elem) or (atom.element is None and atom.name == name)) and not hasMatch[i] and len(atom.bondedTo) == len(bondedTo[position]) and atom.externalBonds == externalBonds[position]:
+            # See if the bonds for this identification are consistent
+
+            allBondsMatch = all((bonded > position or matches[bonded] in atom.bondedTo for bonded in bondedTo[position]))
+            if allBondsMatch:
+                # This is a possible match, so trying matching the rest of the residue.
+
+                matches[position] = i
+                hasMatch[i] = True
+                if _findAtomMatches(atoms, template, bondedTo, externalBonds, matches, hasMatch, position+1):
+                    return True
+                hasMatch[i] = False
+    return False
+
+def _match_template_and_res_name(topology, residue_map, forcefield):
```

Comment:
Note that the template signatures matching onlymeans that you have the same composition of atoms; you can still subsequently fail matching in the `_findAtomMatches()` stage because the bonds don't match up.